README
NAME

    Do

ABSTRACT

    Modern Perl

SYNOPSIS

      package User;
    
      use Do 'Class';
    
      has 'fname';
      has 'lname';
    
      method greet(Str $name) {
        my $fname = $self->fname;
        my $lname = $self->lname;
    
        "Hey $name, I'm $fname $lname";
      }
    
      package main;
    
      my $user = User->new(fname => 'Al', lname => 'Newkirk');
    
      say $user->greet('You');
    
      1;

DESCRIPTION

    This package aims to provide a modern Perl development framework and
    foundational set of types, functions, classes, patterns, and interfaces
    for jump-starting application development. This package inherits all
    behavior from Data::Object.

SLOGAN

    If you're doing something modern with Perl, start here!

PURPOSE

    This package provides a framework for modern Perl development,
    embracing Perl's multi-paradigm programming nature, flexibility and
    vast ecosystem that many of engineers already know and love. The power
    of this framework comes from the extendable (yet fully optional) type
    library which is integrated into the object system and
    type-constrainable subroutine signatures (supporting functions, methods
    and method modifiers). We also provide classes which wrap Perl 5 native
    data types and provides methods for operating on the data.

CONVENTION

    Contrary to the opinion of some, modern Perl programming can be
    extremely well-structured and beautiful, leveraging many advanced
    concepts found in other languages, and some which aren't. Abilities
    like method modification also referred to as augmentation, reflection,
    advanced object-orientation, type-constrainable object attributes,
    type-constrainable subroutine signatures (with named and positional
    arguments), as well roles (similar to mixins or interfaces in other
    languages). This framework aims to serve as an entrypoint to leveraging
    those abilities.

      use Do;

    The "Do" package is an alias and subclass of this package. It
    encapsulates all of the framework's features, is minimalist, and is
    meant to be the first import in a new class or module.

      use Data::Object;

    Both import statements are funcationally equivalent, enable the same
    functionality, and can be configured equally. This is what's enabled
    whenever you import the "Do" or "Data::Object" package into your
    namespace.

      # basics
      use strict;
      use warnings;
    
      # loads say, state, switch, etc
      use feature ':5.14';
    
      # loads type constraints
      use Data::Object::Library;
    
      # loads function/method signatures
      use Data::Object::Signatures;
    
      # imports keywords and super "do" function, etc
      use Data::Object::Export;
    
      # enables method calls on native data types
      use Data::Object::Autobox;

    To explain by way of example: The following established a user-defined
    type library where user-defined classes, roles, etc, will be
    automatically registered.

      package App;
    
      use Do 'Library';
    
      1;

    The following creates a class representing a user which has the ability
    to greet another person. This class is type-library aware and will
    register itself as a type constraint.

      package App::User;
    
      use Do 'Class', 'App';
    
      has name => (
        is  => 'ro',
        isa => 'Str',
        req => 1
      );
    
      method hello(AppUser $user) {
        return 'Hello '. $user->name .'. How are you?';
      }
    
      1;

    The following is a script which is type-library aware that creates a
    function that returns how one user greets another user.

      package main;
    
      use App::User;
    
      use Do 'Core', 'App';
    
      fun greetings(AppUser $u1, AppUser $u2) {
        return $u1->hello($u2);
      }
    
      my $u1 = User->new(name => 'Jane');
      my $u2 = User->new(name => 'June');
    
      say(greetings($u1, $u2)); # Hello June ...

    This demonstrates much of the power of this framework in one simple
    example. If you're new to Perl, the code above creates a class with a
    single (read-only string) attribute called name and a single method
    called hello, then registers the class in a user-defined type-library
    called App where all user-defined type constraints will be stored and
    retrieved (and reified). The main program (namespace) initializes the
    framework and specifies the user-defined type library to use in the
    creation of a single function greetings which takes two arguments which
    must both be instances of the class we just created. Please see
    Data::Object for more information and usages.

CONFIGURATION

    It's important to note that while the example showcases much of what's
    possible with this framework, all of the sophistication is totally
    optional. For example, method and function signatures are optionally
    typed, so the declarations would work just as well without the types
    specified. In fact, you could then remove the App type library
    declarations and even resort rewriting the method and function as
    plain-old Perl subroutines. This flexibility to be able to enable more
    advanced capabilities is common in the Perl ecosystem and is one of the
    things we love most. The wiring-up of things! If you're familiar with
    Perl, this framework is in-part the wiring up of Moo (with Moose
    support), Type::Tiny, Function::Parameters, Try::Tiny and data objects
    in a cooperative and cohesive way that feels like it's native to the
    language. Please feel free to use as much or as little of the framework
    as you need and are comfortable with.

INSTALLATION

    If you have cpanm, you only need one line:

      $ cpanm -qn Do

    If you don't have cpanm, get it! It takes less than a minute,
    otherwise:

      $ curl -L https://cpanmin.us | perl - -qn Do

    Add Do to the list of dependencies in cpanfile:

      requires "Do" => "1.60"; # 1.60 or newer

    If cpanm doesn't have permission to install modules in the current Perl
    installation, it will automatically set up and install to a local::lib
    in your home directory. See the local::lib documentation for details on
    enabling it in your environment. We recommend using a Perlbrew
    <https://github.com/gugod/app-perlbrew> or Plenv
    <https://github.com/tokuhirom/plenv> environment. These tools will help
    you manage multiple Perl installations in your $HOME directory. They
    are completely isolated Perl installations.

CREDITS

    Al Newkirk, +294

    Anthony Brummett, +10

    José Joaquín Atria, +1

AUTHOR

    Al Newkirk, awncorp@cpan.org

LICENSE

    Copyright (C) 2011-2019, Al Newkirk, et al.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.

PROJECT

    GitHub <https://github.com/iamalnewkirk/do>

    Projects <https://github.com/iamalnewkirk/do/projects>

    Milestones <https://github.com/iamalnewkirk/do/milestones>

    Contributing
    <https://github.com/iamalnewkirk/do/blob/master/CONTRIBUTE.mkdn>

    Issues <https://github.com/iamalnewkirk/do/issues>

SEE ALSO

    To get the most out of this distribution, consider reading the
    following:

    Do

    Data::Object

    Data::Object::Class

    Data::Object::ClassHas

    Data::Object::Role

    Data::Object::RoleHas

    Data::Object::Library



CHANGES
Changelog for Do

1.80 2019-09-19T12:03:50
 - Added release updates
 - Updated release script
 - Fixed trailing-comma issue in D/O/R/Dumpable
 - Updated recommended version
 - Ignores POD-only files in generation
 - Cleanup Keywords, Add New Keywords (#293)
 - Adding pretty_dump, pretty_print, and pretty_say to
   Data::Object::Role::Dumpable (#294)
 - Implemented D/O/Boolean

1.76 2019-09-16T05:14:40
 - Built release version 1.76
 - Added release updates
 - Implemented D/O/String/Func/Render (#287)
 - Updated GH workflows and release script (#286)

1.75 2019-09-15T22:32:02
 - Built release version 1.75
 - Added release updates
 - Fixed functor constraint declaration errors (typos)
 - Fixed functor constraints; Allows raw or obj values (#273)
 - Added NumberLike constraint to type library (#272)
 - Updated GH workflows (#270)
 - Fix snakecase method for strings (#261) (#264)

1.74 2019-09-14T18:31:47
 - Built release version 1.74
 - Added release updates
 - Fix snakecase method for strings (#261) (#264)
 - Updated Documentation and Software License (#266)
 - Cleaned-up Core Roles (#263)
 - Implemented D/O/Try,/R/Tryable (#259)
 - Implemented D/O/Name (#257)
 - Updated GH templates and workflows (#249)

1.70 2019-09-09T00:38:07
 - Built release version 1.70
 - Added release updates
 - Implemented D/O/Args, Opts, Vars, and Cli (#246)
 - Added defaults for data objects (#242)
 - Update main module POD (#240)
 - Implemented D/O/Vars (#238)

1.60 2019-09-06T20:08:10
 - Built release version 1.60
 - Added release updates
 - Fixed release process and testing workflow (#234)
 - Framework Housekeeping (#233)
 - Updated GH templates and workflows (#232)
 - Updated testing.yml
 - Built release version 1.09
 - Fixed leaked token in release workflow
 - Documented inheritance in POD (#219)
 - Fixed failed POD generation for routines (#218)
 - Miscellaneous Housekeeping (#204)
 - Updated issue templates
 - Updated GitHub actions and issue template (#196)
 - Build 1.07 (#192)
 - Updated GitHub action workflows (#190)
 - Updated GitHub action workflows (#189)
 - Build 1.05 (#187)
 - Build 1.03 (#184)
 - Released v1.01 (#181)
 - Performance optimizations (#179)
 - Build 0.98 (#177)
 - Build 0.97 (#176)
 - Build 0.96 (#173)
 - Build 0.95 (#172)
 - Revert "Build 0.94 (#170)" (#171)
 - Build 0.94 (#170)
 - Build 0.93 (#168)
 - Build 0.92 (#167)
 - Build 0.92 (#166)
 - Build 0.91 (#158)
 - Build 0.90 (#157)
 - Build 0.80
 - Build 0.70
 - Update dist.ini to generate a MYMETA.yml
 - Built release version 0.61
 - Tidy source and update build scripts (#15)
 - Built release version 0.60
 - Reformatted module files
 - Reformatted test files
 - Updated to work with new Type::Tiny::Signatures release
 - Updated build.
 - Added global types array to library.
 - Added the Proxyable role to support proxy methods and autoloading.
 - Modified POD in SYNOPSIS for existing roles.
 - Added support for alternate Data::Object namespacing.
 - v0.58
 - Updated SEE ALSO POD.
 - Updated POD for Hash values method.
 - Modify the Hash fold method to support empty hashes.
 - v0.57
 - Updated codification examples for Array, Hash, and Code packages.
 - v0.56
 - Updated POD.
 - Updated codification implementation.
 - Added reify function.
 - Refactored codification to to use Sub::Quote.
 - Updated main module synopsis.
 - v0.55
 - Updated POD for Syntax packages and removed dead code.
 - v0.54
 - Updated the .travis.yml file to include Perl versions 5.14+.
 - Fixed code object dump method test failure.
 - v0.53
 - Fixed Type::Tiny::XS type constraint test failure.
 - v0.52
 - Updated POD for all syntax packages.
 - v0.51
 - Updated and cleaned-up POD across all files.
 - v0.50
 - v0.42
 - Updated gitignore.
 - Complete refactoring, massive changes and cleanup.
 - Miscellaneous changes.
 - Imported Data::Object::Signatures into the codebase.
 - Imported Data::Object::Library into the codebase.
 - Updated dist.ini, .gitignore, and POD for Data::Object, etc.
 - Added method and function signatures to type classes.
 - Miscellaneous changes and cleanup.
 - Added support for producing immutable type objects.
 - Make all data type objects throwable.
 - Refactor import statements and foreign function calls.
 - Added Perl 5.14 and Type::Tiny(::Signatues) dependencies.
 - Updated POD for Data::Object.
 - Updated the distribution description.
 - Updated the POD for the new method on each class.
 - Removed the Constructor role.
 - Updated the POD for the syntax packages.
 - v0.41
 - Updated POD for syntax packages.
 - v0.40
 - Changed behaviour for req and opt functions.
 - Updated POD for const.
 - v0.39
 - Added req and opt functions to the syntax packages.
 - Removed the alias function.
 - Added tests and POD for the const function.
 - Refactored the constructor for all type classes.
 - v0.38
 - Updated miscellaneous role descriptions.
 - Updated POD for Role.
 - Updated the POD for the Singleton class.
 - Updated the ABSTRACT for all relevant classes.
 - Update the Data::Object description.
 - Added const function to the core export tag.
 - v0.37
 - Added the const function, Updated the load and alias functions.
 - Updated POD for teh Exception class.
 - Updated the main SYNOPSIS.
 - Modified constructor for all objects.
 - v0.36
 - Added SEE ALSO sections to the class and role packages.
 - v0.35
 - Fixed and documented the defaulter behavior.
 - v0.34
 - Updated the alias function POD.
 - v0.33
 - Fixed attribute altering and defaulter behavior.
 - Updated the SEE ALSO documentation.
 - Added alias function.
 - v0.32
 - Added "alt", "def", and "defaulter" functions to the syntax packages.
 - v0.31
 - Fixed and simplified all class/object constructors.
 - v0.30
 - Fixed documentation for Class, Role, and Exception packages.
 - v0.29
 - Added documentation for Class, Role, and Exception packages.
 - Added tests for the Syntax DSL packages.
 - Updated Syntax package descriptions.
 - v0.28
 - Generating documentation for the Syntax DSL packages.

Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Modern Perl",
  "AUTHOR" => "Al Newkirk <awncorp\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Do",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.014000",
  "NAME" => "Do",
  "PREREQ_PM" => {
    "Function::Parameters" => "2.001003",
    "Import::Into" => "1.002005",
    "Moo" => "2.000001",
    "Readonly" => "2.00",
    "Role::Tiny" => "2.000006",
    "Try::Tiny" => "0.30",
    "Type::Tiny" => "1.000005",
    "autobox" => "2.82"
  },
  "TEST_REQUIRES" => {
    "Function::Parameters" => "2.001003",
    "Import::Into" => "1.002005",
    "Moo" => "2.000001",
    "Readonly" => "2.00",
    "Role::Tiny" => "2.000006",
    "Try::Tiny" => "0.30",
    "Type::Tiny" => "1.000005",
    "autobox" => "2.82"
  },
  "VERSION" => "1.75",
  "test" => {
    "TESTS" => "t/0.01/data/*.t t/0.01/data/object/*.t t/0.01/data/object/array/*.t t/0.01/data/object/code/*.t t/0.01/data/object/float/*.t t/0.01/data/object/hash/*.t t/0.01/data/object/integer/*.t t/0.01/data/object/library/*.t t/0.01/data/object/number/*.t t/0.01/data/object/regexp/*.t t/0.01/data/object/scalar/*.t t/0.01/data/object/signatures/*.t t/0.01/data/object/string/*.t t/0.01/data/object/undef/*.t t/0.50/data/*.t t/0.50/data/object/*.t t/0.50/data/object/class/*.t t/0.50/data/object/func/*.t t/0.50/data/object/regexp/*.t t/0.90/*.t t/0.90/can/*.t t/0.90/use/*.t"
  }
);


my %FallbackPrereqs = (
  "Function::Parameters" => "2.001003",
  "Import::Into" => "1.002005",
  "Moo" => "2.000001",
  "Readonly" => "2.00",
  "Role::Tiny" => "2.000006",
  "Try::Tiny" => "0.30",
  "Type::Tiny" => "1.000005",
  "autobox" => "2.82"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
